(dev) Oauth troubleshooting (#25564)
DataDog/documentation: content/en/developers/integrations/oauth_for_integrations.md

For more information on Datadog's OAuth implementation, see the [Datadog OAuth2 documentation][1].

> Publishing an OAuth client does not result in a published integration. Your integration only appears on the [Integrations page][16] after you've completed a separate publication process. For information on creating and publishing an integration, see [Build and integration][18].

Reviewer: ```suggestion
Publishing an OAuth client does not result in a published integration. Your integration only appears on the [Integrations page][16] after you've completed a separate publication process. For information on creating and publishing an integration, see [Build an integration][18].
```

   4. Go through the OAuth flow and authorize your integration.

#### Create an API Key

If your OAuth client requests the `api_keys_write` scope, ensure that you can successfully make a request to the `marketplace` endpoint with your token in the headers of the request. 

If successful, this request returns an API key that you can find on the [API Keys Management page][10]. You must securely save this key to use it for submitting data into Datadog on behalf of the user. **You cannot access this API key value again after the initial request response**.

#### Test multiple Datadog sites

> You cannot test other organizations with your testing client, but you can test your OAuth client can work across multiple [Datadog sites][8] by copying your client into your EU sandbox and kicking off authorization.

Reviewer: ```suggestion
You cannot test other organizations with your testing client, but you can verify that your OAuth client works across multiple [Datadog sites][8] by copying your client into your EU sandbox and kicking off authorization.
```

   1. If you do not have access to a sandbox account on a different site, contact `ecosystems@datadog.com`.
   2. Export your app manifest from the organization in the *original* US1 Datadog site by navigating to the app you've created in the Developer Platform, clicking the Gear icon to the right of **Documentation**, and clicking **Export App Manifest**.
   3. In your EU sandbox organization, navigate to the Developer Platform and import your app manifest from Step 2.
   4. After successfully importing your manifest, navigate to the **OAuth & Permissions** tab to find your OAuth client, along with its client ID and client secret. Update your OAuth implementation to use these credentials.
   5. Navigate to the **Test Authorization** button, click it, and go through the OAuth flow.

After your OAuth client is published, you can test freely from other organizations.

##### Cross-regional support

> To make OAuth work for users across all Datadog regions, you need to ensure that you're making the correct API calls based on a user's region. When the user kicks off authorization from the Datadog tile, a site parameter is sent on redirect from the onboarding url. You use this site parameter in your calls to the authorization and token endpoints.

Reviewer: ```suggestion
To make OAuth work for users across all Datadog regions, you need to ensure that you're making the correct API calls based on a user's region. When the user kicks off authorization from the Datadog tile, a site parameter is sent on redirect from the onboarding URL. You use this site parameter in your calls to the authorization and token endpoints.
```

If a user kicks off authorization directly from your platform, this site parameter is not sent and the user is instead prompted to select their site on the Datadog authorization page.

> Make sure you test calls to the Datadog API that match the user's region. For example, `https://trace.browser-intake-datadoghq.com` for US, and `https://public-trace-http-intake.logs.datadoghq.eu` for EU.

Reviewer 2: We can link out. I think it's clear how to change it

> Make sure you test calls to the Datadog API that match the user's region. For example, `https://trace.browser-intake-datadoghq.com` for US, and `https://public-trace-http-intake.logs.datadoghq.eu` for EU.

Reviewer 2: https://docs.datadoghq.com/getting_started/site/

> Make sure you test calls to the Datadog API that match the user's region. For example, `https://trace.browser-intake-datadoghq.com` for US, and `https://public-trace-http-intake.logs.datadoghq.eu` for EU.

Author: Added a link to the networking page below. That one has a lot more examples (and I'm about to start some work to update it with a bunch more).

   4. Go through the OAuth flow and authorize your integration.

#### Create an API Key

> If your OAuth client requests the `api_keys_write` scope, ensure that you can successfully make a request to the `marketplace` endpoint with your token in the headers of the request. 

Reviewer 2: Can we link the marketplace endpoint to here: https://docs.datadoghq.com/developers/authorization/oauth2_endpoints/?tab=apikeycreationendpoints

For more information about publishing your integration tile and creating your pull request, see the [Marketplace and Integrations documentation][7].

## Troubleshooting

> ### The list of API scopes does not include sending metrics, events and logs

Reviewer: ```suggestion
### The list of API scopes does not include sending metrics, events, and logs
```

To send data to Datadog, use the `api_keys_write` scope when generating an API key on behalf of the user. For more information, see [Create an API key](#create-an-api-key).

### API key and token expiration

Refresh tokens do not expire unless the user revokes authorization or the partner revokes the token. If the partner revokes the token, the user must reauthorize the integration to generate new refresh and access tokens. For more information, see the [OAuth2 Authorization Endpoints Reference][13].

### Retrieving API keys

> After you create an key using the [api_keys/marketplace][14] endpoint, the key is returned in the response. The key cannot be regenerated or viewed again. Ensure you store the key securely for continuous data transmission. If you lose your API key, follow these steps to revoke and recreate it:

Reviewer: ```suggestion
After you create a key using the [api_keys/marketplace][14] endpoint, the key is returned in the response. The key cannot be regenerated or viewed again. Ensure you store the key securely for continuous data transmission. If you lose your API key, follow these steps to revoke and recreate it:
```

> 

Reviewer 2: can we add a note under the steps: 
Note: The steps above applies only to partners testing in their own partner sandbox account

Author: I added that bit to the section title to make this clearer

1. Navigate to the [Datadog API Keys Management page][15].
1. Look for the API key named `OAuth Client API Key` and select it.
1. Click **Revoke** to disable the API key.
1. Follow the steps in [Create an API key](#create-an-api-key) to create a new key.
1. Reinstall the integration and repeat the OAuth flow.

### Invalid client ID

Error
: `invalid_request - Invalid client_id parameter value`

Until an OAuth client is published, you can only authorize the client from the account it was created in (the partner's sandbox account). This error occurs if you try to authorize the client outside of that account before the client is published.

### Forbidden errors related to app keys

> 

Reviewer 2: The error is: 
{"errors":["Forbidden"]}

> 

Author: Added the error and combined made the two sections that deal with forbidden errors subsections of this one.

OAuth clients use an `access_token` for authentication. Use the `access_token` to make calls to Datadog API endpoints by sending it as a part of the authorization header of your request:

```python
headers = {"Authorization": "Bearer {}".format(access_token)}
```

For more information, see [Implement the OAuth protocol][17].

### Forbidden errors on API requests

> If you're getting a forbidden error when trying to make an API call to a specific endpoint and you've enabled the correct scope for that endpoint, it's possible your API key, session, or OAuth token is invalid or has expired.

Author: Question about this one. Is the issue here that the token has been revoked? We mention above that that's the only way a token can expire. So perhaps we should combine the two sections?

> If you're getting a forbidden error when trying to make an API call to a specific endpoint and you've enabled the correct scope for that endpoint, it's possible your API key, session, or OAuth token is invalid or has expired.

Reviewer 2: True, if the token is revoked, then this will cause all the above.

> If you're getting a forbidden error when trying to make an API call to a specific endpoint and you've enabled the correct scope for that endpoint, it's possible your API key, session, or OAuth token is invalid or has expired.

Reviewer 3: Let's merge them then.

### API errors when making a call with subdomains

> When connecting to the Datadog API, do not include the subdomain in the API call. For example, use `datadoghq.eu` instead of `bigcorp.datadoghq.eu`.

Reviewer 3: @emarsha94 I went back through our emails and found this (subject line: [Datadog] Update: Re: OAuth2 beta): 'We have a customer, LayerX, who is based in Japan and is using Datadog with a custom subdomain: [layerx-corp-prd.datadoghq.com](http://layerx-corp-prd.datadoghq.com/). They managed to install the Datadog integration into their workspace, which set up a webhook subscription that tries to send events to a URL with their subdomain: https://http-intake.logs.layerx-corp-prd.datadoghq.com/api/v2/logs?ddsource=notion.

However, whenever we try to send an event to the URL we get an SSL cert error:
Hostname/IP does not match certificate's altnames: Host: [http-intake.logs.layerx-corp-prd.datadoghq.com](http://http-intake.logs.layerx-corp-prd.datadoghq.com/). is not in the cert's altnames: DNS:*.[datadoghq.com](http://datadoghq.com/), DNS:[datadoghq.com](http://datadoghq.com/)'. 

I believe this is where this FAQ came from. Can we repurpose the error from that email here to show an example?

Reviewer 2: I don't think this is always the error which is why I don't think we should put one here. For example, the SSO cases where they are using the sub-domain instead of the site.

Author: This is perhaps something we can iterate on if we find frequent patterns that lead back to this issue

Author: Added the error to this one

For more information about publishing your integration tile and creating your pull request, see the [Marketplace and Integrations documentation][7].

## Troubleshooting

### The list of API scopes does not include sending metrics, events, and logs

To send data to Datadog, use the `api_keys_write` scope when generating an API key on behalf of the user. For more information, see [Create an API key](#create-an-api-key).

> ### API key and token expiration
> 
> Refresh tokens do not expire unless the user revokes authorization or the partner revokes the token. If the partner revokes the token, the user must reauthorize the integration to generate new refresh and access tokens. For more information, see the [OAuth2 Authorization Endpoints Reference][13].
> 
> ### Retrieving API keys in your partner sandbox account
> 
> After you create a key using the [api_keys/marketplace][14] endpoint, the key is returned in the response. The key cannot be regenerated or viewed again. Ensure you store the key securely for continuous data transmission. If you lose your API key, follow these steps to revoke and recreate it:
> 
> 1. Navigate to the [Datadog API Keys Management page][15].
> 1. Look for the API key named `OAuth Client API Key` and select it.
> 1. Click **Revoke** to disable the API key.
> 1. Follow the steps in [Create an API key](#create-an-api-key) to create a new key.
> 1. Reinstall the integration and repeat the OAuth flow.

Reviewer 2: ```suggestion
```

### Invalid client ID

Error
: `invalid_request - Invalid client_id parameter value`

Until an OAuth client is published, you can only authorize the client from the account it was created in (the partner's sandbox account). This error occurs if you try to authorize the client outside of that account before the client is published.

### Forbidden errors

Error
: `{"errors":["Forbidden"]}`

This error might be related to an app key or an issue with your API authentication credentials.

#### App key use

OAuth clients use an `access_token` for authentication. Use the `access_token` to make calls to Datadog API endpoints by sending it as a part of the authorization header of your request:

```python
headers = {"Authorization": "Bearer {}".format(access_token)}
```

For more information, see [Implement the OAuth protocol][17].

#### API requests

> If you're getting a forbidden error when trying to make an API call to a specific endpoint and you've enabled the correct scope for that endpoint, it's possible your API key, session, or OAuth token is invalid or has expired.

Reviewer 2: ```suggestion
If you're getting a forbidden error when trying to make an API call to a specific endpoint and you've enabled the correct scope for that endpoint, it's possible your API key, session, or OAuth token is invalid or has expired.

### API key and token expiration

Refresh tokens do not expire unless the user revokes authorization or the partner revokes the token. If the partner revokes the token, the user must reauthorize the integration to generate new refresh and access tokens. For more information, see the [OAuth2 Authorization Endpoints Reference][13].

### Retrieving API keys in your partner sandbox account

After you create a key using the [api_keys/marketplace][14] endpoint, the key is returned in the response. The key cannot be regenerated or viewed again. Ensure you store the key securely for continuous data transmission. If you lose your API key, follow these steps to revoke and recreate it:

1. Navigate to the [Datadog API Keys Management page][15].
1. Look for the API key named `OAuth Client API Key` and select it.
1. Click **Revoke** to disable the API key.
1. Follow the steps in [Create an API key](#create-an-api-key) to create a new key.
1. Reinstall the integration and repeat the OAuth flow.

```
